File: zh-hant/codes/csharp/chapter_greedy/coin_change_greedy.cs
﻿/**
* File: coin_change_greedy.cs
* Created Time: 2023-07-21
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_greedy;

public class coin_change_greedy {
    /* 零錢兌換：貪婪 */
    int CoinChangeGreedy(int[] coins, int amt) {
        // 假設 coins 串列有序
        int i = coins.Length - 1;
        int count = 0;
        // 迴圈進行貪婪選擇，直到無剩餘金額
        while (amt > 0) {
            // 找到小於且最接近剩餘金額的硬幣
            while (i > 0 && coins[i] > amt) {
                i--;
            }
            // 選擇 coins[i]
            amt -= coins[i];
            count++;
        }
        // 若未找到可行方案，則返回 -1
        return amt == 0 ? count : -1;
    }

    [Test]
    public void Test() {
        // 貪婪：能夠保證找到全域性最優解
        int[] coins = [1, 5, 10, 20, 50, 100];
        int amt = 186;
        int res = CoinChangeGreedy(coins, amt);
        Console.WriteLine("\ncoins = " + coins.PrintList() + ", amt = " + amt);
        Console.WriteLine("湊到 " + amt + " 所需的最少硬幣數量為 " + res);

        // 貪婪：無法保證找到全域性最優解
        coins = [1, 20, 50];
        amt = 60;
        res = CoinChangeGreedy(coins, amt);
        Console.WriteLine("\ncoins = " + coins.PrintList() + ", amt = " + amt);
        Console.WriteLine("湊到 " + amt + " 所需的最少硬幣數量為 " + res);
        Console.WriteLine("實際上需要的最少數量為 3 ，即 20 + 20 + 20");

        // 貪婪：無法保證找到全域性最優解
        coins = [1, 49, 50];
        amt = 98;
        res = CoinChangeGreedy(coins, amt);
        Console.WriteLine("\ncoins = " + coins.PrintList() + ", amt = " + amt);
        Console.WriteLine("湊到 " + amt + " 所需的最少硬幣數量為 " + res);
        Console.WriteLine("實際上需要的最少數量為 2 ，即 49 + 49");
    }
}

File: ja/codes/java/chapter_greedy/coin_change_greedy.java
/**
 * File: coin_change_greedy.java
 * Created Time: 2023-07-20
 * Author: krahets (krahets@163.com)
 */

package chapter_greedy;

import java.util.Arrays;

public class coin_change_greedy {
    /* コイン交換：貪欲法 */
    static int coinChangeGreedy(int[] coins, int amt) {
        // coins リストはソート済みと仮定する
        int i = coins.length - 1;
        int count = 0;
        // 残額がなくなるまで貪欲選択を繰り返す
        while (amt > 0) {
            // 残額以下で最も近い硬貨を見つける
            while (i > 0 && coins[i] > amt) {
                i--;
            }
            // coins[i] を選択する
            amt -= coins[i];
            count++;
        }
        // 実行可能な解が見つからなければ -1 を返す
        return amt == 0 ? count : -1;
    }

    public static void main(String[] args) {
        // 貪欲法：大域最適解を保証できる
        int[] coins = { 1, 5, 10, 20, 50, 100 };
        int amt = 186;
        int res = coinChangeGreedy(coins, amt);
        System.out.println("\ncoins = " + Arrays.toString(coins) + ", amt = " + amt);
        System.out.println("合計 " + amt + " に必要な最小硬貨枚数は " + res);

        // 貪欲法：大域最適解を保証できない
        coins = new int[] { 1, 20, 50 };
        amt = 60;
        res = coinChangeGreedy(coins, amt);
        System.out.println("\ncoins = " + Arrays.toString(coins) + ", amt = " + amt);
        System.out.println("合計 " + amt + " に必要な最小硬貨枚数は " + res);
        System.out.println("実際に必要な最小枚数は 3、つまり 20 + 20 + 20");

        // 貪欲法：大域最適解を保証できない
        coins = new int[] { 1, 49, 50 };
        amt = 98;
        res = coinChangeGreedy(coins, amt);
        System.out.println("\ncoins = " + Arrays.toString(coins) + ", amt = " + amt);
        System.out.println("合計 " + amt + " に必要な最小硬貨枚数は " + res);
        System.out.println("実際に必要な最小枚数は 2、つまり 49 + 49");
    }
}
